libkbfs/block_retrieval_worker.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"io"
+	"reflect"
+
+	"golang.org/x/net/context"
+)
+
+// blockRetrievalWorker processes blockRetrievalQueue requests
+type blockRetrievalWorker struct {
+	blockGetter
+	stopCh chan struct{}
+	queue  *blockRetrievalQueue
+}
+
+// run runs the worker loop until Shutdown is called
+func (brw *blockRetrievalWorker) run() {
+	for {
+		select {
+		case <-brw.stopCh:
+			return
+		default:
+			brw.HandleRequest()
+		}
+	}
+}
+
+// newBlockRetrievalWorker returns a blockRetrievalWorker for a given
+// blockRetrievalQueue, using the passed in blockGetter to obtain blocks for
+// requests.
+func newBlockRetrievalWorker(bg blockGetter, q *blockRetrievalQueue) *blockRetrievalWorker {
+	brw := &blockRetrievalWorker{
+		blockGetter: bg,
+		stopCh:      make(chan struct{}),
+		queue:       q,
+	}
+	go brw.run()
+	return brw
+}
+
+// notifyBlockRequestor copies the source block into the request's block
+// pointer, and notifies the channel that the request is waiting on. Should be
+// called in a goroutine.
+func notifyBlockRequestor(req *blockRetrievalRequest, source reflect.Value, err error) {
+	// Copy the decrypted block to the caller
+	dest := reflect.ValueOf(req.block).Elem()
```

Comment:
It's unfortunate that we have to use reflection a lot in this PR. That comes from the `Get` function, which takes a `Block` as an output parameter. It's been a low-priority TODO to fix that to take an enum (file block vs. dir block) and return a `Block` object of the correct type. Fixing that is a yak, and probably not worth doing yet, but I think we can at least make these new files do that internally, and just translate to the old API at the boundary (i.e., in `BlockOpsStandard.Get`), so that the tech. debt doesn't increase.

In particular, I think it suffices to:

- Make a block type enum;
- Make `Request` take a `blockType`, and instead of the returned channel returning error, have it return a `{ Block, error }` struct, or a `{ *FileBlock, *DirBlock, error }` struct, where exactly one field will be filled in.
- `Request` can then double-check the passed-in `blockType` with the one in the existing request, if any.
- in `BlockOpsStandard.Get`, check the type of the passed-in `Block` and use it to fill in a `blockType` appropriately. Then when the channel returns, then assign to the passed-in `Block` appropriately.

Some tests IIRC define a separate `testBlock` type, which you'd have to fix up. But that should be easy enough, I think.